os/sys/stack-check.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2017, University of Bristol - http://www.bris.ac.uk/
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+/**
+ * \addtogroup stack
+ * @{
+ */
+
+/**
+ * \file
+ *     Implementation of the stack checker library.
+ * \author
+ *     Atis Elsts <atis.elsts@bristol.ac.uk>
+ */
+
+#include "contiki.h"
+#include "sys/stack-check.h"
+#include "dev/watchdog.h"
+#include <string.h>
+
+#include "sys/log.h"
+#define LOG_MODULE "Stack"
+#define LOG_LEVEL LOG_LEVEL_MAIN
+
+/*---------------------------------------------------------------------------*/
+#if STACK_CHECK_ENABLED
```

Comment:
Do we _really_ need this? Won't the linker do the job for us? Would it be an idea to initialise the library conditionally in `main()` and then always compile the code here?

Reply:
Yes, that would work. Do you prefer having the `#if` guards in `main()`, or have something else in mind?

Reply:
No problem with `#if` guards in `main()`, we use them for orchestra and the shell too.

Reply:
Done, I like that there's no need for multiple "versions" of the functions in the stack check library code now!